packages/vaadin-virtual-list/src/virtualizer-iron-list-adapter.js

```diff
@@ -0,0 +1,432 @@
+        this._vidxOffset = 0;
+      } else {
+        targetVirtualIndex = OFFSET_ADJUST_MIN_THRESHOLD;
+        this._vidxOffset = index - targetVirtualIndex;
+      }
+    } else {
+      this._vidxOffset = index - targetVirtualIndex;
+    }
+
+    this.__skipNextVirtualIndexAdjust = true;
+    super.scrollToIndex(targetVirtualIndex);
+    if (this.firstVisibleIndex !== index - this._vidxOffset && this._scrollTop < this._maxScrollTop && !this.grid) {
+      // Second invocation to scrollToIndex may be needed to workaround an issue in iron-list
+      super.scrollToIndex(targetVirtualIndex);
+    }
+    this._scrollHandler();
+  }
+
+  flush() {
+    this._resizeHandler();
+    flush();
+    this._scrollHandler();
+    this.__scrollReorderDebouncer && this.__scrollReorderDebouncer.flush();
+    this.__debouncerWheelAnimationFrame && this.__debouncerWheelAnimationFrame.flush();
+  }
+
+  update(startIndex = 0, endIndex = this.size - 1) {
+    this.__getVisibleElements().forEach((el) => {
+      if (el.__virtualIndex >= startIndex && el.__virtualIndex <= endIndex) {
+        this.updateElement(el, el.__virtualIndex);
+      }
+    });
+  }
+
+  __getIndexScrollOffset(index) {
+    const element = this.__getVisibleElements().find((el) => el.__virtualIndex === index);
+    return element ? this.scrollTarget.getBoundingClientRect().top - element.getBoundingClientRect().top : undefined;
+  }
+
+  set size(size) {
+    if (size === this.size) {
+      return;
+    }
+
+    let fvi = this.firstVisibleIndex + this._vidxOffset;
+
+    const fviOffsetBefore = this.__getIndexScrollOffset(fvi);
+
+    this.__size = size;
+    this._itemsChanged({
+      path: 'items'
+    });
+
+    fvi = Math.min(fvi, size - 1);
+    this.scrollToIndex(fvi);
+
+    const fviOffsetAfter = this.__getIndexScrollOffset(fvi);
+    if (fviOffsetBefore !== undefined && fviOffsetAfter !== undefined) {
+      this._scrollTop += fviOffsetBefore - fviOffsetAfter;
+    }
+
+    if (!this.elementsContainer.children.length) {
+      requestAnimationFrame(() => this._resizeHandler());
+    }
+  }
+
+  get size() {
+    return this.__size;
+  }
+
+  /** @private */
+  get _scrollTop() {
+    return this.scrollTarget.scrollTop;
+  }
+
+  /** @private */
+  set _scrollTop(top) {
+    this.scrollTarget.scrollTop = top;
+  }
+
+  /** @private */
+  get items() {
+    return {
+      length: Math.min(this.size, MAX_VIRTUAL_COUNT)
+    };
+  }
+
+  /** @private */
+  get offsetHeight() {
+    return this.scrollTarget.offsetHeight;
+  }
+
+  /** @private */
+  get $() {
+    return {
+      items: this.scrollContainer
+    };
+  }
+
+  /** @private */
+  updateViewportBoundaries() {
+    const styles = window.getComputedStyle(this.scrollTarget);
+    this._scrollerPaddingTop = this.scrollTarget === this ? 0 : parseInt(styles['padding-top'], 10);
+    this._isRTL = Boolean(styles.direction === 'rtl');
+    this._viewportWidth = this.elementsContainer.offsetWidth;
```

Comment:
I wonder if there is a difference between properties starting from `__` and `_`? What is the rule behind this?

Reply:
It's a convention described [here](https://polymer-library.polymer-project.org/2.0/docs/devguide/properties#private-and-protected-properties). IIRC the convention was adjusted between Polymer versions 1 and 2 so there may be some mismatches (private members that haven't yet been updated to use double-undescore).

Reply:
Got it, thanks!